addons/binding/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/UPBBridgeHandler.java

```diff
@@ -0,0 +1,192 @@
+import org.openhab.binding.upb.internal.UPBReader;
+import org.openhab.binding.upb.internal.UPBWriter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gnu.io.CommPortIdentifier;
+import gnu.io.NoSuchPortException;
+import gnu.io.PortInUseException;
+import gnu.io.SerialPort;
+import gnu.io.UnsupportedCommOperationException;
+
+/**
+ * Handler for Universal Powerline Bus (UPB) that reads and writes messages to
+ * and from the UPB modem.
+ *
+ * @author Chris Van Orman
+ * @since 1.9.0
+ */
+public class UPBBridgeHandler extends BaseBridgeHandler implements UPBMessageSender, UPBReader.Listener {
+
+    private static final String CONFIG_NETWORK = "network";
+    private static final String CONFIG_PORT = "serialPort";
+
+    private static final Logger logger = LoggerFactory.getLogger(UPBBridgeHandler.class);
+
+    private String port;
+    private byte network = 0;
+    private SerialPort serialPort;
+    private List<UPBMessageListener> listeners = new CopyOnWriteArrayList<>();
+    private UPBReader upbReader;
+    private UPBWriter upbWriter;
+
+    /**
+     * Instantiates a new {@link UPBBridgeHandler}.
+     * 
+     * @param bridge the bridge to be handled
+     */
+    public UPBBridgeHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    @Override
+    public void initialize() {
+
+        // Read necessary config values
+        this.network = ((Number) getConfig().get(CONFIG_NETWORK)).byteValue();
+        this.port = (String) getConfig().get(CONFIG_PORT);
+
+        logger.debug("Parsed UPB configuration:");
+        logger.debug("Serial port: {}", port);
+        logger.debug("UPB Network: {}", network & 0xff);
+
+        try {
+            serialPort = openSerialPort();
+            upbReader = new UPBReader(new DataInputStream(serialPort.getInputStream()));
+            upbWriter = new UPBWriter(new DataOutputStream(serialPort.getOutputStream()), upbReader);
+
+            upbReader.addListener(this);
+            updateStatus(ThingStatus.ONLINE);
+        } catch (Exception e) {
+            logger.error("Error opening serial port [{}].", port, e);
+
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Error opening serial port.");
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (upbReader != null) {
+            upbReader.shutdown();
+        }
+
+        if (upbWriter != null) {
+            upbWriter.shutdown();
+        }
+
+        if (serialPort != null) {
+            logger.debug("Closing serial port");
+            serialPort.close();
+        }
+    }
+
+    private SerialPort openSerialPort() {
+        SerialPort serialPort = null;
+        CommPortIdentifier portId;
+        try {
+            portId = CommPortIdentifier.getPortIdentifier(port);
+        } catch (NoSuchPortException e1) {
+            String ports = StringUtils.join(findAvailablePorts(), ", ");
+            throw new RuntimeException(
```

Comment:
Better not throw a raw `RuntimeException` better choose a specific one that applies or create a custom one.